core/src/main/scala/org/apache/spark/rdd/UnionRDD.scala

```diff
@@ -62,8 +62,14 @@ class UnionRDD[T: ClassTag](
     var rdds: Seq[RDD[T]])
   extends RDD[T](sc, Nil) {  // Nil since we implement getDependencies
 
+  // visible for testing
+  private[spark] val isPartitionEvalParallel: Boolean =
+    rdds.length > conf.getInt("spark.rdd.parallelListingThreshold", 10)
+
   override def getPartitions: Array[Partition] = {
-    val array = new Array[Partition](rdds.map(_.partitions.length).sum)
+    val parRDDs = if (isPartitionEvalParallel) rdds.par else rdds
```

Comment:
Yeah we had this discussion above. The downside is that you make a pool of threads that sits around and/or has to spin up just for this block. I agree with the potential problem, though lots of things fall apart if user code jams the shared thread pool. Since this isn't that resource or time sensitive, seemed good to do the simple thing here, but I don't feel strongly about it.